appcypher-syntax.md

```diff
@@ -0,0 +1,947 @@
+
+    ```swift
+    map(filter(map(arr, (x) => x * 2), (x) => x > 2), (x) => x * 2)
+    ```
+
+    ```swift
+    arr
+        .map((x) => x * 2)
+        .filter((x) => x > 2)
+        .map((x) => x * 2)
+    ```
+
+
+
+## TRAITS
+- Ratio does not support inheritance. Instead it choses composition-based type extension.
+
+    ```rust
+    trait Callable {
+        call: (...Int) -> Int
+    }
+    ```
+
+    ```swift
+    impl Sum: Callable {
+        func call(args: ...Int) -> Int {
+            var result = 0
+            for i in args {
+                result += i
+            }
+            result
+        }
+    }
+    ```
+
+## UNSAFE BLOCK
+- The unsafe block is necessary for scoping the usage of unsafe/unstable features like pointer arithmetic and destructive casting.
+
+    ```rust
+    let base_ptr = some_ref as *Int
+
+    let base_ptr = unsafe {
+        base_ptr.add(20)
+    }
+    ```
+
+
+## CASTING
+- Non-destructive casting uses `as` keyword and it is only supported for scalar types.
+
+    ```rust
+    let val = 1000 as f32
+    ```
+
+- For destructive casting, compiler-intrinsic `transmute` function can be used.
+
+    ```rust
+    let s = "String"
+    let buf = unsafe { transmute:[String, Buffer](s) }
+    ```
+
+## ACCESS MODIFIERS
+- Everything is private by default, but you can make member of a module or type public using the `pub` keyword.
+
+    ```rs
+    pub type Person {
+        pub name,
+        age
+    }
+    ```
+
+    ```swift
+    pub func new(name, age) -> Self {
+        Self {
+            name, age
+        }
+    }
+    ```
+
+
+## IMPORT
+- Ratio is a module-based language. To use things from other modules, you have to import from them.
+
+    ```py
+    import .dir {foo}
+    ```
+
+- Ratio has a special file called `index.ratio`. It is used to expose modules in a particular folder. Modules of a folder can only be accessed if the folder has an `index.ratio` that exposes them.
+
+    ```py
+    import .dir
+    ```
+
+    If there is a folder called `dir`, this import can only be valid if there is a file path `./dir/index.ratio`
+
+    If there is no folder called `dir`, then there has to be a file `./dir.ratio`
+
+
+- Ratio has two kinds of imports.
+
+    #### Regular relative import
+
+    The imported module is relative to the source file defining the import. The import symbol is preceded by a dot.
+
+    ```py
+    import .dir
+    ```
+
+    `dir` here can either be `./dir.ratio` or `./dir/index.ratio`
+
+    #### Root relative import
+
+    The imported module is relative to the project's root folder or it is some external dependency.
+
+    ```go
+    import dir
+    ```
+
+    `dir` here can either be `src/dir.ratio` or `src/dir/index.ratio` or some external dependency named `dir` with a path `src/index.ratio`.
+
+- Ratio supports aliasing an imported element '
+
+    ```py
+    import ext {foo}
+    import .module.some {foo as bar}
+
+    foo()
+    bar()
```

Comment:
For the import, we should only use the last approach and eliminate the use of `.` before a module path. Any module imported should be looked up in the relative folder, then other paths where the language module can be such as the project root folder.

Reply:
This is going to cause unintended behavior. I'm afraid, Python got it a bit right here. 

If the import resolver goes from relative to root in it's search, it is going to result in unwanted behavior where you actually want it to choose a module from root rather than relatively.
 
This will inadvertently lead to having to rename modules, just to get import resolver to work right.